x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/rule/RuleExecutor.java

```diff
@@ -21,6 +21,25 @@
 public abstract class RuleExecutor<TreeType extends Node<TreeType>> {
 
     private final Logger log = LogManager.getLogger(getClass());
+    /**
+     * Sub-logger intended to show only changes made by the rules. Intended for debugging.
+     *
+     * Enable it like this for the respective optimizers and the analyzer, resp. all inheritors of this class:
+     * <pre>{@code
+     * PUT localhost:9200/_cluster/settings
+     *
+     * {
+     *   "transient" : {
+     *     "logger.org.elasticsearch.xpack.esql.analysis.Analyzer.changes": "TRACE",
+     *     "logger.org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizer.changes": "TRACE",
+     *     "logger.org.elasticsearch.xpack.esql.optimizer.LocalLogicalPlanOptimizer.changes": "TRACE",
+     *     "logger.org.elasticsearch.xpack.esql.optimizer.PhysicalPlanOptimizer.changes": "TRACE",
+     *     "logger.org.elasticsearch.xpack.esql.optimizer.LocalPhysicalPlanOptimizer.changes": "TRACE"
+     *   }
+     * }
+     * }</pre>
+     */
+    private final Logger changeLog = LogManager.getLogger(getClass().getName() + ".changes");
```

Comment:
You tested this locally, right? I'm pretty sure this is right, but I can't write code without some kind of mistake.